ApeWorX/ape
# feat: adhoc network support

### What I did



fixes: #598 

### How I did it

* Handle when user puts URL at end of ecosystem / network choice string, e.g.:
```
--network ethereum:mainnet:https://foo.bar
--network ::https://foo.bar
```
* Handle when user omits ecosystem and network entirely and use "adhoc"-like settings (geth provider reliant)
```
--network https://foo.bar
```

Here are examples:

Using ecosystem and network:

```
(angel39) ➜  ape-demo-project git:(main) ✗ ape console --network ethereum:mainnet:https://<omit>:<omit>@erigon<omit>/
INFO: Connecting to existing Erigon node at '<omit>'

In [1]: provider
Out[1]: <geth chain_id=1>

In [2]: provider.chain_id
Out[2]: 1

In [3]: provider.get_block("latest")
Out[3]: Block(num_transactions=135, hash=HexBytes('0xda1eb42477d29acdbd2e3bd2c6a517d66b687bc12a63efaca2bda8a39a3fa7ef'), number=15219795, parent_hash=HexBytes('0xf64ef0ffc7843277a29621b186673572b03bd3429cb9420744e427bcc9201979'), size=25401, timestamp=1658861362, gas_limit=29970705, gas_used=5695467, base_fee=16623710745, difficulty=11766754802216255, total_difficulty=54914317311543363044806)
```

And an example without using ecosystem or network:

```
(angel39) ➜  ape-demo-project git:(main) ✗ ape console --network https://rpc.ftm.tools
WARNING: Connecting Geth plugin to non-Geth client 'go-opera'.

In [1]: provider.chain_id
Out[1]: 250

In [2]: provider
Out[2]: <geth chain_id=250>

In [3]: provider.get_block("latest")
Out[3]: Block(num_transactions=6, hash=HexBytes('0x000218e9000011dcfa1c7fadf646fb78f357b9bdaee67ab14725721b616f498d'), number=43606522, parent_hash=HexBytes('0x000218e9000011cfa764771321513d1d92e8d3b7839b25b70169a836e24a0101'), size=14145, timestamp=1658861132, gas_limit=281474976710655, gas_used=13969043, base_fee=1000000000, difficulty=0, total_difficulty=0)
```

### How to verify it

See examples above!
Basically, try to connect to any random EVM-like network.
Anything that would work with a HTTPProvider in web3 should work this way as well.

### Checklist
